File: crates/breez-sdk/bindings/examples/cli/langs/python/src/breez_cli/issuer.py
import argparse

from breez_sdk_spark import (
    BurnIssuerTokenRequest,
    CreateIssuerTokenRequest,
    FreezeIssuerTokenRequest,
    MintIssuerTokenRequest,
    UnfreezeIssuerTokenRequest,
)

from breez_cli.serialization import print_value

# Issuer subcommand names (used for REPL completion)
ISSUER_COMMAND_NAMES = [
    "issuer token-balance",
    "issuer token-metadata",
    "issuer create-token",
    "issuer mint-token",
    "issuer burn-token",
    "issuer freeze-token",
    "issuer unfreeze-token",
]


def _parser(name, description=""):
    return argparse.ArgumentParser(prog=f"issuer {name}", description=description)


# --- token-balance ---

def _build_token_balance_parser():
    return _parser("token-balance", "Get issuer token balance")

async def _handle_token_balance(token_issuer, _args):
    result = await token_issuer.get_issuer_token_balance()
    print_value(result)


# --- token-metadata ---

def _build_token_metadata_parser():
    return _parser("token-metadata", "Get issuer token metadata")

async def _handle_token_metadata(token_issuer, _args):
    result = await token_issuer.get_issuer_token_metadata()
    print_value(result)


# --- create-token ---

def _build_create_token_parser():
    p = _parser("create-token", "Create a new issuer token")
    p.add_argument("name", help="Name of the token")
    p.add_argument("ticker", help="Ticker symbol")
    p.add_argument("decimals", type=int, help="Number of decimal places")
    p.add_argument("-f", "--is-freezable", action="store_true", default=False,
                   help="Whether the token is freezable")
    p.add_argument("max_supply", type=int, help="Maximum supply")
    return p

async def _handle_create_token(token_issuer, args):
    result = await token_issuer.create_issuer_token(
        CreateIssuerTokenRequest(
            name=args.name,
            ticker=args.ticker,
            decimals=args.decimals,
            is_freezable=args.is_freezable,
            max_supply=args.max_supply,
        )
    )
    print_value(result)


# --- mint-token ---

def _build_mint_token_parser():
    p = _parser("mint-token", "Mint supply of the issuer token")
    p.add_argument("amount", type=int, help="Amount to mint")
    return p

async def _handle_mint_token(token_issuer, args):
    result = await token_issuer.mint_issuer_token(
        MintIssuerTokenRequest(amount=args.amount)
    )
    print_value(result)


# --- burn-token ---

def _build_burn_token_parser():
    p = _parser("burn-token", "Burn supply of the issuer token")
    p.add_argument("amount", type=int, help="Amount to burn")
    return p

async def _handle_burn_token(token_issuer, args):
    result = await token_issuer.burn_issuer_token(
        BurnIssuerTokenRequest(amount=args.amount)
    )
    print_value(result)


# --- freeze-token ---

def _build_freeze_token_parser():
    p = _parser("freeze-token", "Freeze tokens at an address")
    p.add_argument("address", help="Address holding the tokens to freeze")
    return p

async def _handle_freeze_token(token_issuer, args):
    result = await token_issuer.freeze_issuer_token(
        FreezeIssuerTokenRequest(address=args.address)
    )
    print_value(result)


# --- unfreeze-token ---

def _build_unfreeze_token_parser():
    p = _parser("unfreeze-token", "Unfreeze tokens at an address")
    p.add_argument("address", help="Address holding the tokens to unfreeze")
    return p

async def _handle_unfreeze_token(token_issuer, args):
    result = await token_issuer.unfreeze_issuer_token(
        UnfreezeIssuerTokenRequest(address=args.address)
    )
    print_value(result)


# ---------------------------------------------------------------------------
# Registry and dispatch
# ---------------------------------------------------------------------------

def _build_issuer_registry():
    return {
        "token-balance": (_build_token_balance_parser(), _handle_token_balance),
        "token-metadata": (_build_token_metadata_parser(), _handle_token_metadata),
        "create-token": (_build_create_token_parser(), _handle_create_token),
        "mint-token": (_build_mint_token_parser(), _handle_mint_token),
        "burn-token": (_build_burn_token_parser(), _handle_burn_token),
        "freeze-token": (_build_freeze_token_parser(), _handle_freeze_token),
        "unfreeze-token": (_build_unfreeze_token_parser(), _handle_unfreeze_token),
    }


_REGISTRY = None

def _get_registry():
    global _REGISTRY
    if _REGISTRY is None:
        _REGISTRY = _build_issuer_registry()
    return _REGISTRY


async def dispatch_issuer_command(args, token_issuer):
    """Dispatch an issuer subcommand given the args after 'issuer'."""
    registry = _get_registry()

    if not args or args[0] == "help":
        print("\nIssuer subcommands:\n")
        for name, (parser, _) in sorted(registry.items()):
            desc = parser.description or ""
            print(f"  issuer {name:30s} {desc}")
        print()
        return

    sub_name = args[0]
    sub_args = args[1:]

    if sub_name not in registry:
        print(f"Unknown issuer subcommand: {sub_name}. Use 'issuer help' for available commands.")
        return

    parser, handler = registry[sub_name]
    try:
        parsed = parser.parse_args(sub_args)
    except SystemExit:
        return

    await handler(token_issuer, parsed)

File: crates/breez-sdk/bindings/examples/cli/langs/dart/lib/issuer.dart
import 'package:args/args.dart';
import 'package:breez_sdk_spark_flutter/breez_sdk_spark.dart';

import 'serialization.dart';

/// Issuer subcommand names (used for help text).
const issuerCommandNames = [
  'issuer token-balance',
  'issuer token-metadata',
  'issuer create-token',
  'issuer mint-token',
  'issuer burn-token',
  'issuer freeze-token',
  'issuer unfreeze-token',
];

typedef IssuerHandler =
    Future<void> Function(TokenIssuer tokenIssuer, List<String> args);

class _IssuerEntry {
  final String description;
  final IssuerHandler handler;
  const _IssuerEntry(this.description, this.handler);
}

Map<String, _IssuerEntry>? _registry;

Map<String, _IssuerEntry> _getRegistry() {
  return _registry ??= {
    'token-balance': _IssuerEntry(
      'Get issuer token balance',
      _handleTokenBalance,
    ),
    'token-metadata': _IssuerEntry(
      'Get issuer token metadata',
      _handleTokenMetadata,
    ),
    'create-token': _IssuerEntry(
      'Create a new issuer token',
      _handleCreateToken,
    ),
    'mint-token': _IssuerEntry(
      'Mint supply of the issuer token',
      _handleMintToken,
    ),
    'burn-token': _IssuerEntry(
      'Burn supply of the issuer token',
      _handleBurnToken,
    ),
    'freeze-token': _IssuerEntry(
      'Freeze tokens at an address',
      _handleFreezeToken,
    ),
    'unfreeze-token': _IssuerEntry(
      'Unfreeze tokens at an address',
      _handleUnfreezeToken,
    ),
  };
}

/// Dispatch an issuer subcommand given the args after 'issuer'.
Future<void> dispatchIssuerCommand(
  List<String> args,
  TokenIssuer tokenIssuer,
) async {
  final registry = _getRegistry();

  if (args.isEmpty || args[0] == 'help') {
    print('\nIssuer subcommands:\n');
    for (final entry
        in registry.entries.toList()..sort((a, b) => a.key.compareTo(b.key))) {
      print('  issuer ${entry.key.padRight(30)} ${entry.value.description}');
    }
    print('');
    return;
  }

  final subName = args[0];
  final subArgs = args.sublist(1);

  if (!registry.containsKey(subName)) {
    print(
      "Unknown issuer subcommand: $subName. Use 'issuer help' for available commands.",
    );
    return;
  }

  await registry[subName]!.handler(tokenIssuer, subArgs);
}

// --- token-balance ---

Future<void> _handleTokenBalance(
  TokenIssuer tokenIssuer,
  List<String> args,
) async {
  final result = await tokenIssuer.getIssuerTokenBalance();
  printValue(result);
}

// --- token-metadata ---

Future<void> _handleTokenMetadata(
  TokenIssuer tokenIssuer,
  List<String> args,
) async {
  final result = await tokenIssuer.getIssuerTokenMetadata();
  printValue(result);
}

// --- create-token ---

Future<void> _handleCreateToken(
  TokenIssuer tokenIssuer,
  List<String> args,
) async {
  final parser =
      ArgParser()..addFlag('is-freezable', abbr: 'f', defaultsTo: false);
  final results = parser.parse(args);

  if (results.rest.length < 4) {
    print(
      'Usage: issuer create-token <name> <ticker> <decimals> <max_supply> [-f]',
    );
    return;
  }
  final name = results.rest[0];
  final ticker = results.rest[1];
  final decimals = int.parse(results.rest[2]);
  final maxSupply = BigInt.parse(results.rest[3]);
  final isFreezable = results.flag('is-freezable');

  final result = await tokenIssuer.createIssuerToken(
    request: CreateIssuerTokenRequest(
      name: name,
      ticker: ticker,
      decimals: decimals,
      isFreezable: isFreezable,
      maxSupply: maxSupply,
    ),
  );
  printValue(result);
}

// --- mint-token ---

Future<void> _handleMintToken(
  TokenIssuer tokenIssuer,
  List<String> args,
) async {
  if (args.isEmpty) {
    print('Usage: issuer mint-token <amount>');
    return;
  }
  final amount = BigInt.parse(args[0]);
  final result = await tokenIssuer.mintIssuerToken(
    request: MintIssuerTokenRequest(amount: amount),
  );
  printValue(result);
}

// --- burn-token ---

Future<void> _handleBurnToken(
  TokenIssuer tokenIssuer,
  List<String> args,
) async {
  if (args.isEmpty) {
    print('Usage: issuer burn-token <amount>');
    return;
  }
  final amount = BigInt.parse(args[0]);
  final result = await tokenIssuer.burnIssuerToken(
    request: BurnIssuerTokenRequest(amount: amount),
  );
  printValue(result);
}

// --- freeze-token ---

Future<void> _handleFreezeToken(
  TokenIssuer tokenIssuer,
  List<String> args,
) async {
  if (args.isEmpty) {
    print('Usage: issuer freeze-token <address>');
    return;
  }
  final address = args[0];
  final result = await tokenIssuer.freezeIssuerToken(
    request: FreezeIssuerTokenRequest(address: address),
  );
  printValue(result);
}

// --- unfreeze-token ---

Future<void> _handleUnfreezeToken(
  TokenIssuer tokenIssuer,
  List<String> args,
) async {
  if (args.isEmpty) {
    print('Usage: issuer unfreeze-token <address>');
    return;
  }
  final address = args[0];
  final result = await tokenIssuer.unfreezeIssuerToken(
    request: UnfreezeIssuerTokenRequest(address: address),
  );
  printValue(result);
}
